src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = {};
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        stateHistory.push({});
+        break;
+
+      case 'addProperties':
+        currentState = { ...(stateHistory.at(-1) || { ...state }) };
+
+        stateHistory.push(Object.assign(currentState, action.extraData));
+        break;
+
+      case 'removeProperties':
+        currentState = { ...(stateHistory.at(-1) || { ...state }) };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        stateHistory.push(currentState);
```

Comment:
Similar to the other cases, push the modified `currentState` outside of the `switch` block after the loop cycle completes.